lib/ibm_cloud_sdk_core/base_service.rb

```diff
@@ -0,0 +1,275 @@
+    @disable_ssl = false
+    @display_name = vars[:display_name]
+
+    if !vars[:iam_access_token].nil? || !vars[:iam_apikey].nil?
+      set_token_manager(iam_apikey: vars[:iam_apikey], iam_access_token: vars[:iam_access_token], iam_url: vars[:iam_url])
+    elsif !vars[:username].nil? && !vars[:password].nil?
+      if vars[:username] == "apikey" && !@icp_prefix
+        iam_apikey(iam_apikey: vars[:password])
+      else
+        @username = vars[:username]
+        @password = vars[:password]
+      end
+    end
+
+    if @display_name && !@username && !@iam_apikey
+      service_name = @display_name.sub(" ", "_").downcase
+      load_from_credential_file(service_name)
+    end
+
+    if vars[:use_vcap_services] && !@username && !@iam_apikey
+      @vcap_service_credentials = load_from_vcap_services(service_name: vars[:vcap_services_name])
+      if !@vcap_service_credentials.nil? && @vcap_service_credentials.instance_of?(Hash)
+        @url = @vcap_service_credentials["url"]
+        @username = @vcap_service_credentials["username"] if @vcap_service_credentials.key?("username")
+        @password = @vcap_service_credentials["password"] if @vcap_service_credentials.key?("password")
+        @iam_apikey = @vcap_service_credentials["iam_apikey"] if @vcap_service_credentials.key?("iam_apikey")
+        @iam_access_token = @vcap_service_credentials["iam_access_token"] if @vcap_service_credentials.key?("iam_access_token")
+        @iam_url = @vcap_service_credentials["iam_url"] if @vcap_service_credentials.key?("iam_url")
+      end
+    end
+
+    raise ArgumentError.new('The username shouldn\'t start or end with curly brackets or quotes. Be sure to remove any {} and \" characters surrounding your username') if check_bad_first_or_last_char(@username)
+    raise ArgumentError.new('The password shouldn\'t start or end with curly brackets or quotes. Be sure to remove any {} and \" characters surrounding your password') if check_bad_first_or_last_char(@password)
+    raise ArgumentError.new('The url shouldn\'t start or end with curly brackets or quotes. Be sure to remove any {} and \" characters surrounding your url') if check_bad_first_or_last_char(@url)
+    raise ArgumentError.new('The apikey shouldn\'t start or end with curly brackets or quotes. Be sure to remove any {} and \" characters surrounding your apikey') if check_bad_first_or_last_char(@iam_apikey)
+
+    @conn = HTTP::Client.new(
+      headers: {}
+    ).use normalize_uri: { normalizer: NORMALIZER }
+  end
+
+  # Initiates the credentials based on the credential file
+  def load_from_credential_file(service_name, separator = "=")
+    credential_file_path = ENV["IBM_CREDENTIALS_FILE"]
+
+    # Home directory
+    if credential_file_path.nil?
+      file_path = ENV["HOME"] + DEFAULT_CREDENTIALS_FILE_NAME
+      credential_file_path = file_path if File.exist?(file_path)
+    end
+
+    # Top-level directory of the project
```

Comment:
What happens when the user specifies a credentials file that exists but it also has one in the home directory? And what about if the one defined doesn't exist? Do we tell him that it doesn't exist?

Reply:
The one specified by user gets precedence even if there is another credentials file in the home directory. And yep, we throw the error message the file doesn't exist.